modules/host/host.go

```diff
@@ -135,47 +423,106 @@ func New(cs modules.ConsensusSet, tpool modules.TransactionPool, wallet modules.
 
 	// Create the host object.
 	h := &Host{
-		cs:     cs,
-		tpool:  tpool,
-		wallet: wallet,
+		cs:           cs,
+		tpool:        tpool,
+		wallet:       wallet,
+		dependencies: dependencies,
 
-		actionItems: make(map[types.BlockHeight]map[types.FileContractID]*contractObligation),
+		// actionItems: make(map[types.BlockHeight]map[types.FileContractID]*contractObligation),
 
-		obligationsByID: make(map[types.FileContractID]*contractObligation),
+		// obligationsByID: make(map[types.FileContractID]*contractObligation),
+
+		lockedStorageObligations: make(map[types.FileContractID]struct{}),
 
 		persistDir: persistDir,
 	}
 
 	// Create the perist directory if it does not yet exist.
-	err := os.MkdirAll(h.persistDir, 0700)
+	err := dependencies.mkdirAll(h.persistDir, 0700)
 	if err != nil {
 		return nil, err
 	}
 
 	// Initialize the logger. Logging should be initialized ASAP, because the
 	// rest of the initialization makes use of the logger.
-	h.log, err = persist.NewLogger(filepath.Join(h.persistDir, logFile))
+	h.log, err = dependencies.newLogger(filepath.Join(h.persistDir, logFile))
+	if err != nil {
+		return nil, err
+	}
+
+	// Open the database containing the host's storage obligation metadata.
+	h.db, err = dependencies.openDatabase(dbMetadata, filepath.Join(h.persistDir, dbFilename))
+	if err != nil {
+		// An error will be returned if the database has the wrong version, but
+		// as of writing there was only one version of the database and all
+		// other databases would be incompatible.
+		_ = h.log.Close()
+		return nil, err
+	}
+	// After opening the database, it must be initalized. Most commonly,
+	// nothing happens. But for new databases, a set of buckets must be
+	// created. Intialization is also a good time to run sanity checks.
+	err = h.initDB()
 	if err != nil {
+		_ = h.log.Close()
+		_ = h.db.Close()
 		return nil, err
 	}
 
 	// Load the prior persistance structures.
 	err = h.load()
 	if err != nil {
+		_ = h.log.Close()
+		_ = h.db.Close()
 		return nil, err
 	}
 
 	// Get the host established on the network.
 	err = h.initNetworking(address)
 	if err != nil {
+		_ = h.log.Close()
+		_ = h.db.Close()
 		return nil, err
 	}
 
 	return h, nil
 }
 
+// New returns an initialized Host.
+func New(cs modules.ConsensusSet, tpool modules.TransactionPool, wallet modules.Wallet, address string, persistDir string) (*Host, error) {
+	return newHost(productionDependencies{}, cs, tpool, wallet, address, persistDir)
+}
+
+// composeErrors will take two errors and compose them into a single errors
```

Comment:
can be more than two.

separating with "and" might be confusing if the error string contains the word "and." Maybe a semicolon would be better.
```go
func compose(errs ...error) error {
	var errStrings []string
	for _, err := range errs {
		if err != nil {
			errStrings = append(errStrings, err.Error())
		}
	}
	return errors.New(strings.Join(errStrings, "; "))
}
```